c10/util/MaybeOwned.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <c10/util/in_place.h>
+
+namespace c10 {
+
+/// A smart pointer around either a borrowed or owned T. Maintains an
+/// internal raw pointer when constructed with borrowed(), with all
+/// the attendant lifetime concerns.  Compare to Rust's
+/// std::borrow::Cow
+/// (https://doc.rust-lang.org/std/borrow/enum.Cow.html).
+template <typename T>
+class TORCH_API MaybeOwned {
+  bool isBorrowed_;
+  union {
+    const T *borrow_;
+    T own_;
+  };
+  explicit MaybeOwned(const T& t) : isBorrowed_(true), borrow_(&t) {}
+  explicit MaybeOwned(T&& t) : isBorrowed_(false), own_(std::move(t)) {}
```

Comment:
should be noexcept if T's move constructor is noexcept (otherwise std::vector and other containers do silly things with this data type and just copies, ignoring the move constructor). Assuming this is supposed to be public of course, otherwise std::vector doesn't really matter.